satellite-placement/satellite.py

```diff
@@ -0,0 +1,69 @@
+"""This example is based on a project by Booz Allen Hamilton researchers
+[#btkrd]. See also link_.
+
+The problem is to choose k sub-constellations of m satellites such that the
+sum of the average coverage within each constellation is maximized.
+
+.. _link: https://www.dwavesys.com/sites/default/files/QuantumForSatellitesQubits-4.pdf
+
+.. [#btkrd] G. Bass, C. Tomlin, V. Kumar, P. Rihaczek, J. Dulny III.
+    Heterogeneous Quantum Computing for Satellite Constellation Optimization:
+    Solving the Weighted K-Clique Problem. 2018 Quantum Sci. Technol. 3 024010.
+    https://arxiv.org/abs/1709.05381
+
+"""
+import itertools
+import random
+
+import dimod
+import neal
+import networkx as nx
+
+# we wish to divide 9 satellites into 3 constellations of 3 satellites each.
+num_satellites = 9
+num_constellations = 3
+
+constellation_size = num_satellites // num_constellations
+
+# don't consider constellations with average score less than score_threshold
+score_threshold = .4
+
+# for this example we will randomly assign a coverage score to each satellite
+coverage = {v: random.uniform(.25, 1) for v in range(num_satellites)}
+
+bqm = dimod.BinaryQuadraticModel.empty(dimod.BINARY)
+
+# first we want to favor combinations with a high score
+for constellation in itertools.combinations(range(num_satellites), constellation_size):
+    # the score is the average coverage for a constellation
+    score = sum(coverage[v] for v in constellation) / len(constellation)
+
+    # to make it smaller, throw out the combinations with a score below
+    # a set threshold
+    if score < score_threshold:
+        continue
+
+    # we subtract the score because we want to minimize the energy
+    bqm.add_variable(frozenset(constellation), -score)
```

Comment:
Because it is a random problem I am ok with this case existing. The user can tweak the `score_threshold` above. I think it would be more unexpected to dynamically choose the threshhold only in a case when the resulting BQM is empty.

Reply:
In any case I avoided the problem by hard-coding the coverages.